File: app/src/main/java/com/tastyradio/playback/Mixer.kt
package com.tastyradio.playback

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.media.AudioFocusRequest
import android.media.AudioManager
import android.os.Handler
import android.os.Looper
import androidx.annotation.OptIn
import androidx.media3.common.AudioAttributes
import androidx.media3.common.C
import androidx.media3.common.MediaItem
import androidx.media3.common.Metadata
import androidx.media3.common.PlaybackException
import androidx.media3.common.Player
import androidx.media3.common.util.UnstableApi
import androidx.media3.datasource.DefaultDataSource
import androidx.media3.datasource.DefaultHttpDataSource
import androidx.media3.exoplayer.DefaultLoadControl
import androidx.media3.exoplayer.ExoPlayer
import androidx.media3.exoplayer.LoadControl
import androidx.media3.exoplayer.source.DefaultMediaSourceFactory
import androidx.media3.extractor.metadata.icy.IcyInfo
import com.tastyradio.data.Station
import androidx.media3.exoplayer.upstream.DefaultLoadErrorHandlingPolicy
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

/**
 * The mixing desk. One [ExoPlayer] per active station, each with its own fader.
 *
 * This is the class the whole app exists for, and the one place where Tasty Radio stops resembling
 * every other radio app: a station is not *the* playback, it's a channel on a mix.
 *
 * Lives for the life of the process (owned by the Application), because both the UI and the
 * playback service talk to the same instance.
 *
 * ## Audio focus
 * There is exactly **one** focus owner: this class. Every player is built with
 * `handleAudioFocus = false`. If each player managed its own focus they would fight — every new
 * player requesting focus would duck or pause the others, and the app's own stations would fade
 * each other out. That bug is invisible with one station and obvious with three.
 */
@OptIn(UnstableApi::class)
class Mixer(private val context: Context) {

    enum class ChannelState { Connecting, Playing, Failed }

    /**
     * One channel of the mix. [fader] is the slider position, 0..1 — not an amplitude.
     *
     * [key] rather than the Room id, because a channel isn't always a saved station: auditioning a
     * search result plays it into the running mix *without* adding it to the collection, and every
     * unsaved station has `id = 0`. Keying on the id would make all auditioned stations collide
     * with each other.
     */
    data class Channel(
        val key: String,
        val station: Station,
        val fader: Float = DEFAULT_FADER,
        val muted: Boolean = false,
        val state: ChannelState = ChannelState.Connecting,
        /** ICY stream metadata — the current track, when the station bothers to send it. */
        val nowPlaying: String? = null,
        val error: String? = null,
    )

    private val main = Handler(Looper.getMainLooper())
    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.Main.immediate)
    private val audioManager = context.getSystemService(AudioManager::class.java)

    /** One per connecting channel: a stream that never arrives has to become a visible failure. */
    private val watchdogs = mutableMapOf<String, Job>()

    /** Insertion-ordered so the mixer rows don't jump around as channels come and go. */
    private val players = LinkedHashMap<String, ExoPlayer>()

    private val _channels = MutableStateFlow<List<Channel>>(emptyList())
    val channels: StateFlow<List<Channel>> = _channels.asStateFlow()

    /** Set while another app holds transient focus and asked us to duck rather than stop. */
    private var ducking = false

    /**
     * Applied to players as they're built. Changing it mid-mix doesn't disturb what's already
     * playing — a new buffer size takes effect the next time a station starts.
     */
    @Volatile var largeBuffer: Boolean = true


    private var focusRequest: AudioFocusRequest? = null
    private var noisyReceiverRegistered = false

    private val focusListener = AudioManager.OnAudioFocusChangeListener { change ->
        when (change) {
            // Live radio can't be resumed from where it left off, so a real loss means stop.
            AudioManager.AUDIOFOCUS_LOSS,
            AudioManager.AUDIOFOCUS_LOSS_TRANSIENT -> stopAll()

            AudioManager.AUDIOFOCUS_LOSS_TRANSIENT_CAN_DUCK -> {
                ducking = true
                applyVolumes()
            }

            AudioManager.AUDIOFOCUS_GAIN -> {
                ducking = false
                applyVolumes()
            }
        }
    }

    /** Headphones pulled out stops the whole mix, from one receiver rather than one per player. */
    private val noisyReceiver = object : BroadcastReceiver() {
        override fun onReceive(context: Context?, intent: Intent?) {
            if (intent?.action == AudioManager.ACTION_AUDIO_BECOMING_NOISY) stopAll()
        }
    }

    // ---------------------------------------------------------------- channel control

    /**
     * Adds a station to the mix. [fader] lets an audition start quietly under what's already
     * playing.
     *
     * @return false if the mix is already full, so the caller can say so.
     */
    fun play(station: Station, fader: Float = DEFAULT_FADER): Boolean {
        val key = station.channelKey()
        if (players.containsKey(key)) return true
        if (players.size >= MAX_CHANNELS) return false

        if (players.isEmpty()) {
            requestFocus()
            registerNoisyReceiver()
            startPlaybackService()
        }

        val player = buildPlayer()
        players[key] = player
        _channels.update { it + Channel(key = key, station = station, fader = fader) }

        player.addListener(ChannelListener(key))
        player.setMediaItem(MediaItem.fromUri(station.streamUrl))
        player.prepare()
        player.playWhenReady = true
        applyVolumes()
        armWatchdog(key)
        return true
    }

    fun stop(key: String) {
        watchdogs.remove(key)?.cancel()
        players.remove(key)?.release()
        _channels.update { list -> list.filterNot { it.key == key } }
        if (players.isEmpty()) releaseFocusAndReceiver()
    }

    fun stopAll() {
        watchdogs.values.forEach { it.cancel() }
        watchdogs.clear()
        players.values.forEach { it.release() }
        players.clear()
        _channels.value = emptyList()
        releaseFocusAndReceiver()
    }

    fun toggle(station: Station): Boolean {
        val key = station.channelKey()
        return if (players.containsKey(key)) {
            stop(key)
            true
        } else {
            play(station)
        }
    }

    fun isLive(station: Station) = players.containsKey(station.channelKey())

    /** Retry a channel that failed — dead stations and flaky ones look identical at first. */
    fun retry(key: String) {
        val channel = _channels.value.firstOrNull { it.key == key } ?: return
        val player = players[key] ?: return
        updateChannel(key) { it.copy(state = ChannelState.Connecting, error = null) }
        player.setMediaItem(MediaItem.fromUri(channel.station.streamUrl))
        player.prepare()
        player.playWhenReady = true
        armWatchdog(key)
    }

    /**
     * A stream that connects but never delivers audio used to sit on `Connecting…` forever, because
     * ExoPlayer keeps retrying a load that isn't erroring — it's just silent. After
     * [STALL_TIMEOUT_MS] the channel becomes an honest failure with a retry button instead.
     */
    private fun armWatchdog(key: String) {
        watchdogs.remove(key)?.cancel()
        watchdogs[key] = scope.launch {
            delay(STALL_TIMEOUT_MS)
            val channel = _channels.value.firstOrNull { it.key == key } ?: return@launch
            if (channel.state == ChannelState.Connecting) {
                players[key]?.stop()
                updateChannel(key) {
                    it.copy(state = ChannelState.Failed, error = "no audio after 20s")
                }
            }
        }
    }

    // ---------------------------------------------------------------- faders

    fun setFader(key: String, fader: Float) {
        updateChannel(key) { it.copy(fader = fader.coerceIn(0f, 1f)) }
        applyVolumes()
    }

    fun setMuted(key: String, muted: Boolean) {
        updateChannel(key) { it.copy(muted = muted) }
        applyVolumes()
    }


    private fun applyVolumes() {
        val duck = if (ducking) DUCK_FACTOR else 1f
        for (channel in _channels.value) {
            val player = players[channel.key] ?: continue
            player.volume = if (channel.muted) 0f else amplitudeFor(channel.fader) * duck
        }
    }

    // ---------------------------------------------------------------- plumbing

    private fun buildPlayer(): ExoPlayer {
        val http = DefaultHttpDataSource.Factory()
            .setUserAgent(USER_AGENT)
            // Radio URLs redirect constantly, including http -> https, which is cross-protocol.
            .setAllowCrossProtocolRedirects(true)
            .setConnectTimeoutMs(15_000)
            .setReadTimeoutMs(15_000)

        val attributes = AudioAttributes.Builder()
            .setUsage(C.USAGE_MEDIA)
            .setContentType(C.AUDIO_CONTENT_TYPE_MUSIC)
            .build()

        val mediaSourceFactory = DefaultMediaSourceFactory(DefaultDataSource.Factory(context, http))
            // Give up after a few attempts so a dead host surfaces as an error rather than an
            // indefinite silent retry loop.
            .setLoadErrorHandlingPolicy(DefaultLoadErrorHandlingPolicy(3))

        return ExoPlayer.Builder(context)
            .setLoadControl(buildLoadControl())
            .setMediaSourceFactory(mediaSourceFactory)
            // false: this Mixer is the single audio-focus owner. See the class comment.
            .setAudioAttributes(attributes, false)
            // Likewise — one central becoming-noisy receiver, not one per player.
            .setHandleAudioBecomingNoisy(false)
            .setWakeMode(C.WAKE_MODE_NETWORK)
            .build()
    }

    /**
     * A deliberately deep buffer. Radio streams on mobile data stutter, and with several playing at
     * once a stall is far more noticeable — one channel dropping out breaks the whole mix. The cost
     * is that playback takes longer to start, which is the right trade for a soundscape you leave
     * running. Roughly a minute of audio when large, ten seconds when not.
     */
    private fun buildLoadControl(): LoadControl {
        val maxMs = if (largeBuffer) 60_000 else 15_000
        val startMs = if (largeBuffer) 5_000 else 1_500
        return DefaultLoadControl.Builder()
            .setBufferDurationsMs(
                /* minBufferMs = */ maxMs,
                /* maxBufferMs = */ maxMs,
                /* bufferForPlaybackMs = */ startMs,
                /* bufferForPlaybackAfterRebufferMs = */ startMs * 2,
            )
            .setBackBuffer(0, false)
            .build()
    }

    private inner class ChannelListener(private val key: String) : Player.Listener {

        override fun onIsPlayingChanged(isPlaying: Boolean) {
            if (isPlaying) {
                watchdogs.remove(key)?.cancel()
                updateChannel(key) { it.copy(state = ChannelState.Playing, error = null) }
            }
        }

        override fun onPlaybackStateChanged(playbackState: Int) {
            if (playbackState == Player.STATE_BUFFERING) {
                updateChannel(key) { it.copy(state = ChannelState.Connecting) }
            }
        }

        /**
         * One dead station must not take the mix down with it — losing a channel of three is
         * survivable, so the failure stays local to its own row.
         */
        override fun onPlayerError(error: PlaybackException) {
            updateChannel(key) {
                it.copy(state = ChannelState.Failed, error = error.errorCodeName)
            }
        }

        override fun onMetadata(metadata: Metadata) {
            for (i in 0 until metadata.length()) {
                val entry = metadata.get(i)
                if (entry is IcyInfo) {
                    val title = entry.title?.trim()?.ifEmpty { null }
                    updateChannel(key) { it.copy(nowPlaying = title) }
                }
            }
        }
    }

    private fun updateChannel(key: String, transform: (Channel) -> Channel) {
        _channels.update { list ->
            list.map { if (it.key == key) transform(it) else it }
        }
    }

    private fun requestFocus() {
        val manager = audioManager ?: return
        val request = AudioFocusRequest.Builder(AudioManager.AUDIOFOCUS_GAIN)
            .setAudioAttributes(
                android.media.AudioAttributes.Builder()
                    .setUsage(android.media.AudioAttributes.USAGE_MEDIA)
                    .setContentType(android.media.AudioAttributes.CONTENT_TYPE_MUSIC)
                    .build()
            )
            .setOnAudioFocusChangeListener(focusListener, main)
            .setWillPauseWhenDucked(false)
            .build()
        focusRequest = request
        manager.requestAudioFocus(request)
    }

    private fun releaseFocusAndReceiver() {
        ducking = false
        focusRequest?.let { audioManager?.abandonAudioFocusRequest(it) }
        focusRequest = null
        if (noisyReceiverRegistered) {
            runCatching { context.unregisterReceiver(noisyReceiver) }
            noisyReceiverRegistered = false
        }
    }

    private fun registerNoisyReceiver() {
        if (noisyReceiverRegistered) return
        context.registerReceiver(noisyReceiver, IntentFilter(AudioManager.ACTION_AUDIO_BECOMING_NOISY))
        noisyReceiverRegistered = true
    }

    /**
     * Playback has to be a foreground service or it gets killed the moment the app is backgrounded.
     * Media3 promotes the service and posts the media notification once the session's player
     * reports playing; this just makes sure the service is alive to be promoted.
     */
    private fun startPlaybackService() {
        runCatching {
            context.startService(Intent(context, SoundscapeService::class.java))
        }
    }

    companion object {
        /**
         * Not a hard limit — past four the returns drop and the costs don't: four streams is
         * ~64 KB/s of sustained mobile data, four decoders, four sockets.
         */
        const val MAX_CHANNELS = 4

        const val DEFAULT_FADER = 0.75f

        private const val DUCK_FACTOR = 0.2f

        /** How long a channel may sit connecting before it's called a failure. */
        private const val STALL_TIMEOUT_MS = 20_000L

        /** radio-browser.info asks for a descriptive User-Agent, and so do some stream hosts. */
        private const val USER_AGENT = "TastyRadio/0.1 (Android; hobby project)"

        /**
         * [ExoPlayer.volume] is linear amplitude, and a fader mapped straight onto it feels broken —
         * everything useful bunches into the top of the travel. Cubing approximates a fader taper:
         * half way up lands around −18 dB, which is where half way up should be.
         */
        fun amplitudeFor(fader: Float): Float {
            val f = fader.coerceIn(0f, 1f)
            return f * f * f
        }

        /** A quiet default, so an audition slips under the running mix instead of over it. */
        const val AUDITION_FADER = 0.4f
    }
}

/**
 * Identifies a channel. Saved stations key on their row id; a station that isn't in the collection
 * — a search result being auditioned — has `id = 0`, so it keys on its stream URL instead.
 */
internal fun Station.channelKey(): String = if (id != 0L) "id:$id" else "url:$streamUrl"

File: web/client/src/audio/mixer.ts
import type HlsType from 'hls.js';
import { ChannelGraph, FLAT, type Tone } from './graph.js';
import type { Station } from '../data/types.js';
import { channelKey } from '../data/types.js';
import { streamUrl, forgetChannel, reportClick } from '../api.js';

/**
 * The mixing desk. One `<audio>` element and one signal path per active station.
 *
 * This is the class the whole app exists for, and the one place where Tasty Radio stops resembling
 * every other radio site: a station is not *the* playback, it's a channel on a mix.
 *
 * ## Why the audio goes through our own server
 * Everything interesting here — per-channel EQ, reverb, metering, recording — needs the stream
 * inside Web Audio. A `MediaElementAudioSourceNode` built over a cross-origin stream without CORS
 * headers is tainted and outputs silence, and essentially no radio station sends those headers. So
 * every stream is relayed same-origin by the proxy. That is not a workaround bolted on afterwards;
 * it is the reason a server exists at all.
 */

export type ChannelState = 'connecting' | 'playing' | 'failed';

export interface Channel {
  /**
   * Identifies the channel. Saved stations key on their id; a station that isn't in the collection
   * — a search result being auditioned — has no id, so it keys on its stream URL instead. Keying on
   * the id would make every auditioned station collide with every other.
   */
  key: string;
  station: Station;
  tone: Tone;
  fader: number;
  muted: boolean;
  state: ChannelState;
  /** ICY stream metadata, pushed from the proxy — the current track, when the station sends one. */
  nowPlaying: string | null;
  error: string | null;
}

interface Live {
  channel: Channel;
  element: HTMLAudioElement;
  graph: ChannelGraph;
  hls: HlsType | null;
  watchdog: number | null;
  meterBuffer: Uint8Array;
}

/**
 * Not a hard limit — past four the returns drop and the costs don't: four streams is a steady
 * ~64 KB/s each way, four decoders, four sockets.
 */
export const MAX_CHANNELS = 4;
export const DEFAULT_FADER = 0.75;
/** A quiet default, so an audition slips under the running mix instead of over it. */
export const AUDITION_FADER = 0.4;

/** How long a channel may sit connecting before it's called a failure. */
const STALL_TIMEOUT_MS = 20_000;

export interface Preset {
  station: Station;
  fader: number;
  muted: boolean;
  tone: Tone;
}

export class Mixer {
  readonly ctx: AudioContext;
  readonly master: GainNode;
  /** What the recorder taps. Post-fader, so a take is exactly what you heard. */
  readonly recordTap: MediaStreamAudioDestinationNode;

  private readonly live = new Map<string, Live>();
  private readonly listeners = new Set<() => void>();
  readonly sid: string;

  /**
   * True when the browser refused to start audio because nobody had interacted with the page yet.
   * The site cues its house mix up on landing, so this is the normal state of a first visit.
   */
  blockedByAutoplay = false;

  constructor() {
    this.ctx = new AudioContext({ latencyHint: 'playback' });
    this.master = this.ctx.createGain();
    this.master.gain.value = 1;
    this.recordTap = this.ctx.createMediaStreamDestination();
    this.master.connect(this.ctx.destination);
    this.master.connect(this.recordTap);

    let sid = sessionStorage.getItem('tastyradio.sid');
    if (!sid) {
      sid = Math.random().toString(36).slice(2) + Date.now().toString(36);
      sessionStorage.setItem('tastyradio.sid', sid);
    }
    this.sid = sid;
    this.listenForMetadata();
    this.wireMediaSession();
  }

  // ---------------------------------------------------------------- observation

  subscribe(listener: () => void): () => void {
    this.listeners.add(listener);
    return () => this.listeners.delete(listener);
  }

  private changed(): void {
    for (const listener of this.listeners) listener();
    this.updateMediaSession();
  }

  get channels(): Channel[] {
    return [...this.live.values()].map((entry) => entry.channel);
  }

  isLive(station: Station): boolean {
    return this.live.has(channelKey(station));
  }

  channel(key: string): Channel | undefined {
    return this.live.get(key)?.channel;
  }

  graphFor(key: string): { graph: ChannelGraph; buffer: Uint8Array } | null {
    const entry = this.live.get(key);
    return entry ? { graph: entry.graph, buffer: entry.meterBuffer } : null;
  }

  // ---------------------------------------------------------------- channel control

  /**
   * Adds a station to the mix. `fader` lets an audition start quietly under what's already playing.
   *
   * @returns false if the mix is already full, so the caller can say so.
   */
  play(station: Station, fader = DEFAULT_FADER, tone: Tone = FLAT): boolean {
    const key = channelKey(station);
    if (this.live.has(key)) return true;
    if (this.live.size >= MAX_CHANNELS) return false;

    // Browsers start the context suspended until a gesture. Every route into here is a click.
    void this.ctx.resume();

    const element = new Audio();
    element.preload = 'none';
    element.autoplay = false;
    // The graph is the only route to the speakers; the element's own output is taken over by
    // createMediaElementSource. Leaving this at 1 keeps the fader the single volume control.
    element.volume = 1;

    const graph = new ChannelGraph(this.ctx, element, this.master);
    const channel: Channel = {
      key,
      station,
      tone,
      fader,
      muted: false,
      state: 'connecting',
      nowPlaying: null,
      error: null,
    };
    const entry: Live = {
      channel,
      element,
      graph,
      hls: null,
      watchdog: null,
      meterBuffer: new Uint8Array(graph.analyser.fftSize),
    };
    this.live.set(key, entry);

    graph.setTone(tone);
    graph.setLevel(fader, false);
    this.attachElementEvents(entry);
    this.start(entry);
    this.changed();

    // Feed radio-browser's popularity signal — the thing search ranks on. Fire and forget.
    if (station.sourceUuid) void reportClick(station.sourceUuid);
    return true;
  }

  private start(entry: Live): void {
    const { channel, element } = entry;
    const source = streamUrl(channel.station.streamUrl, this.sid, channel.key);

    entry.hls?.destroy();
    entry.hls = null;

    // Pause before repointing, then load once.
    //
    // ⚠️ Do **not** clear `src` and call `load()` in between. That runs the resource selection
    // algorithm against an empty source, which fires `error` with MEDIA_ERR_SRC_NOT_SUPPORTED — a
    // spurious "no supported source was found" on a stream that is perfectly fine. It cost an
    // evening: RadCap kept dying on the landing mix while the proxy was demonstrably serving it.
    element.pause();

    // Most radio is a plain audio stream. HLS is a real but small minority of the corpus, and
    // hls.js is most of this app's JavaScript, so it is fetched only when something actually needs
    // it rather than by everybody on every visit.
    const looksHls = /\.m3u8(\?|$)/i.test(channel.station.streamUrl);
    const nativeHls = element.canPlayType('application/vnd.apple.mpegurl') !== '';

    if (looksHls && !nativeHls) {
      void this.startHls(entry, source);
    } else {
      element.src = source;
      element.load();
      this.beginPlayback(entry);
    }
    this.armWatchdog(entry);
  }

  private async startHls(entry: Live, source: string): Promise<void> {
    let Hls: typeof HlsType;
    try {
      Hls = (await import('hls.js')).default;
    } catch {
      this.fail(entry, 'could not load the HLS player');
      return;
    }
    // The channel may have been stopped while that was downloading.
    if (!this.live.has(entry.channel.key)) return;

    if (!Hls.isSupported()) {
      this.fail(entry, 'this browser cannot play HLS');
      return;
    }

    const hls = new Hls({ enableWorker: true, lowLatencyMode: false });
    entry.hls = hls;
    hls.on(Hls.Events.ERROR, (_event, data) => {
      if (!data.fatal) return;
      this.fail(entry, data.details ?? 'HLS error');
    });
    hls.loadSource(source);
    hls.attachMedia(entry.element);
    this.beginPlayback(entry);
  }

  private beginPlayback(entry: Live): void {
    void entry.element.play().catch((error: DOMException) => {
      // A rejected play() before any user gesture is the browser's autoplay policy, not a dead
      // station. It is emphatically *not* a failure — the watchdog is stood down and the UI offers
      // a tap, because calling this channel broken would be a lie about a stream we never tried.
      if (error.name === 'NotAllowedError') {
        this.clearWatchdog(entry);
        this.blockedByAutoplay = true;
        this.changed();
        return;
      }
      this.fail(entry, error.message);
    });
  }

  /**
   * Starts everything that the autoplay policy refused. **Must be called from a real user gesture**
   * — that is the entire point of the policy, and a resume() outside one is ignored.
   */
  async unblock(): Promise<void> {
    await this.ctx.resume().catch(() => undefined);
    this.blockedByAutoplay = false;
    for (const entry of this.live.values()) {
      // A full restart rather than another play(): a stream can drop while it sits waiting for
      // somebody to tap, and resuming a dead element hands back a channel that is already broken.
      this.update(entry.channel.key, (channel) => ({ ...channel, state: 'connecting', error: null }));
      this.start(entry);
    }
    this.changed();
  }

  private attachElementEvents(entry: Live): void {
    const { element } = entry;

    element.addEventListener('playing', () => {
      this.clearWatchdog(entry);
      this.update(entry.channel.key, (channel) => ({ ...channel, state: 'playing', error: null }));
    });
    element.addEventListener('waiting', () => {
      this.update(entry.channel.key, (channel) =>
        channel.state === 'failed' ? channel : { ...channel, state: 'connecting' },
      );
    });
    element.addEventListener('stalled', () => this.armWatchdog(entry));
    element.addEventListener('error', () => {
      // hls.js reports its own failures; the element's error during an HLS load is noise.
      if (entry.hls) return;
      // An element with nothing to play is being torn down, not failing.
      if (!element.getAttribute('src')) return;
      this.fail(entry, describeMediaError(element.error));
    });
    // A live stream that "ends" has dropped the connection.
    element.addEventListener('ended', () => this.fail(entry, 'stream ended'));
  }

  /**
   * A stream that connects but never delivers audio would otherwise sit on "connecting" forever,
   * because the browser keeps waiting on a load that isn't erroring — it's just silent. After
   * twenty seconds the channel becomes an honest failure with a retry button instead.
   */
  private armWatchdog(entry: Live): void {
    this.clearWatchdog(entry);
    // Nothing is loading while the autoplay policy holds us; counting down would only produce a
    // false failure on a stream that was never given a chance.
    if (this.blockedByAutoplay) return;
    entry.watchdog = window.setTimeout(() => {
      if (entry.channel.state === 'connecting') this.fail(entry, 'no audio after 20s');
    }, STALL_TIMEOUT_MS);
  }

  private clearWatchdog(entry: Live): void {
    if (entry.watchdog !== null) {
      clearTimeout(entry.watchdog);
      entry.watchdog = null;
    }
  }

  /** One dead station must not take the mix down with it, so a failure stays local to its own row. */
  private fail(entry: Live, reason: string): void {
    this.clearWatchdog(entry);
    this.update(entry.channel.key, (channel) => ({ ...channel, state: 'failed', error: reason }));
  }

  /** Retry a channel that failed — dead stations and flaky ones look identical at first. */
  retry(key: string): void {
    const entry = this.live.get(key);
    if (!entry) return;
    this.update(key, (channel) => ({ ...channel, state: 'connecting', error: null }));
    this.start(entry);
  }

  stop(key: string): void {
    const entry = this.live.get(key);
    if (!entry) return;
    this.clearWatchdog(entry);
    entry.hls?.destroy();
    entry.element.pause();
    // Pointing at nothing is what actually closes the socket; pause alone leaves it open.
    entry.element.removeAttribute('src');
    entry.element.load();
    entry.graph.dispose();
    this.live.delete(key);
    void forgetChannel(this.sid, key);
    this.changed();
  }

  stopAll(): void {
    for (const key of [...this.live.keys()]) this.stop(key);
  }

  toggle(station: Station): boolean {
    const key = channelKey(station);
    if (this.live.has(key)) {
      this.stop(key);
      return true;
    }
    return this.play(station);
  }

  /**
   * Replaces the whole mix in one go. Loading a saved soundscape is "play this instead of what's
   * on", not "add to it" — otherwise recalling a mix while something plays gives you neither.
   */
  load(presets: Preset[]): void {
    this.stopAll();
    for (const preset of presets.slice(0, MAX_CHANNELS)) {
      if (!this.play(preset.station, preset.fader, preset.tone)) continue;
      if (preset.muted) this.setMuted(channelKey(preset.station), true);
    }
  }

  // ---------------------------------------------------------------- faders

  setFader(key: string, fader: number): void {
    const entry = this.live.get(key);
    if (!entry) return;
    const clamped = Math.min(1, Math.max(0, fader));
    entry.channel.fader = clamped;
    entry.graph.setLevel(clamped, entry.channel.muted);
    this.changed();
  }

  setMuted(key: string, muted: boolean): void {
    const entry = this.live.get(key);
    if (!entry) return;
    entry.channel.muted = muted;
    entry.graph.setLevel(entry.channel.fader, muted);
    this.changed();
  }

  setTone(key: string, tone: Tone): void {
    const entry = this.live.get(key);
    if (!entry) return;
    entry.channel.tone = tone;
    entry.graph.setTone(tone);
    this.changed();
  }

  /** Solo is cheap on top of mute and genuinely useful when balancing. */
  soloOnly(key: string): void {
    const others = [...this.live.values()].filter((entry) => entry.channel.key !== key);
    const alreadySoloed = others.every((entry) => entry.channel.muted);
    for (const entry of others) this.setMuted(entry.channel.key, !alreadySoloed);
    this.setMuted(key, false);
  }

  private update(key: string, transform: (channel: Channel) => Channel): void {
    const entry = this.live.get(key);
    if (!entry) return;
    entry.channel = transform(entry.channel);
    this.changed();
  }

  // ---------------------------------------------------------------- now playing

  /**
   * ICY metadata arrives interleaved in the audio bytes, which only the proxy can see, so titles
   * come back over one SSE connection rather than being read here.
   */
  private listenForMetadata(): void {
    const events = new EventSource(`/api/events?sid=${encodeURIComponent(this.sid)}`);
    events.onmessage = (message) => {
      try {
        const { channel, title } = JSON.parse(message.data) as { channel: string; title: string };
        this.update(channel, (existing) => ({ ...existing, nowPlaying: title || null }));
      } catch {
        /* a malformed event is not worth a crash */
      }
    };
  }

  // ---------------------------------------------------------------- one session for the mix

  /**
   * The lock-screen and headset controls should treat the soundscape as a single thing — pressing
   * pause on a headset shouldn't have to choose a station. Same reasoning as the Android app's
   * single `MediaSession` over the aggregate.
   */
  private wireMediaSession(): void {
    if (!('mediaSession' in navigator)) return;
    try {
      navigator.mediaSession.setActionHandler('pause', () => this.stopAll());
      navigator.mediaSession.setActionHandler('stop', () => this.stopAll());
    } catch {
      /* older browsers reject unknown actions */
    }
  }

  private updateMediaSession(): void {
    if (!('mediaSession' in navigator)) return;
    const channels = this.channels;
    if (channels.length === 0) {
      navigator.mediaSession.playbackState = 'none';
      navigator.mediaSession.metadata = null;
      return;
    }
    const names = channels.map((channel) => channel.station.name);
    navigator.mediaSession.playbackState = 'playing';
    navigator.mediaSession.metadata = new MediaMetadata({
      title: channels.length === 1 ? names[0] : `${channels.length} stations`,
      artist: names.join(' · '),
      album: 'Tasty Radio',
    });
  }
}

function describeMediaError(error: MediaError | null): string {
  if (!error) return 'playback failed';
  switch (error.code) {
    case MediaError.MEDIA_ERR_ABORTED:
      return 'aborted';
    case MediaError.MEDIA_ERR_NETWORK:
      return 'network dropped';
    case MediaError.MEDIA_ERR_DECODE:
      return 'could not decode this stream';
    case MediaError.MEDIA_ERR_SRC_NOT_SUPPORTED:
      return 'stream format not supported';
    default:
      return 'playback failed';
  }
}
